### PR TITLE
Add encoding parameter, allow parameters to be omitted if they are ().

### PR DESCRIPTION
**Pull request by BranLwyd**
I do this by creating non-parameterized versions of the traits, then
creating blanket implementations allowing the non-parameterized trait
to act as the parameterized trait with a parameter of ().
Implementations should implement the non-parameterized version of the
trait if they don't need a parameter.

The encoding parameter will practically be used in some PPM messages to
allow the required authentication key to be used as part of the encoding
logic.

### CODE REVIEW

File: src/codec.rs
Comment (tgeoghegan) [inline src/codec.rs:82]: Why is `Sized` needed? Neither method o n the trait takes `self` by value or deals in `Self`.

File: src/vdaf/suite.rs
Comment (tgeoghegan) [inline src/vdaf/suite.rs:6]: Changes in this file will collide with https://github.com/abetterinternet/libprio-rs/pull/181

**Review by tgeoghegan — APPROVED** [review]
This is great and will save us so much `&()` noise.

File: src/codec.rs
Comment (BranLwyd) [reply src/codec.rs:82]: Without it, compilation fails with the following error:
```
error[E0599]: the method `get_encoded_with_param` exists for reference `&Self`, but its trait bounds were not satisfied
  --> src/codec.rs:88:14
   |
88 |         self.get_encoded_with_param(&())
   |              ^^^^^^^^^^^^^^^^^^^^^^ method cannot be called on `&Self` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `Self: Sized`
           which is required by `Self: ParameterizedEncode<()>`
           `&Self: Encode`
           which is required by `&Self: ParameterizedEncode<()>`
```


I think this was because of the implicit `Sized` bound in the blanket implementation; I added an explicit `?Sized` and compilation works out without specifying that `Encode` requires `Sized`.

Comment (tgeoghegan) [reply src/codec.rs:82]: In any case I don't think requiring `Sized` is a problem here.

File: src/vdaf/suite.rs
Comment (BranLwyd) [reply src/vdaf/suite.rs:6]: It looks like `suite.rs` is removed by that PR, but it looks like trivial updates will be require to the `prg.rs` that replaces it (a `Decode` implementation will need to drop its `<()>` parameterization & unused `_decoding_parameter` parameter).

Want me to wait for that PR to make it in to merge this PR? I'd like to get this merged sooner rather than later to unblock work on `janus` but I think I could do some local development by overriding the crate location in my local `cargo.toml`.

Comment (tgeoghegan) [reply src/vdaf/suite.rs:6]: Your call. I think Chris is likely to merge his change later today.

Comment (BranLwyd) [reply src/vdaf/suite.rs:6]: I waited, the merge wasn't too bad & probably easier for me to handle than to foist off on Chris.

**Pull request merged**
